rfcs/0000-local-overlay-store.md

```diff
@@ -0,0 +1,408 @@
+I.e. a lower layer of store dir store object must have lower store metadata (exact mechanism is abstract and unspecified, could be SQlite DB or daemon), and an upper layer of store object must have an overlay DB entry.
+However, when we just have the store object in the lower layer, we may also have metadata in the upper layer.
+That means there are two cases when the metadata is in both layers: the duplicated case (both layers filesystem) and deduplicated case (just lower layer filesystem).
+
+## Operation
+
+As discussed in the motivation, store objects are never knowingly duplicated.
+The `local-overlay` store while in operation ensures that store objects are stored exactly once:
+either in the lower store or the upper layer directory.
+No file system data should ever be duplicated by `local-overlay` itself.
+
+Non-filesystem data, what goes in the DB (references, signatures, etc.) is duplicated.
+Any store object from the lower store that the `local-overlay` needs has that information copied into the `local-overlay` store's DB.
+This includes information for the closure of any such store object, because the normal closure property enforced by the DB's foreign key constraints is upheld.
+
+Store objects can still end up duplicated if the lower store later gains a store object the upper store already had.
+This is because when the `local-overlay` is remounted, it doesn't know how the lower store may have changed, and when the lower store is added to, any upper store directories / DBs are not in general visible either.
+We can have a "fsck" operation however that manually scans for missing / duplicated objects.
+
+## Read-only `local` Store
+
+*This was [already approved](https://github.com/NixOS/nix/pull/8356#event-9483342493) by the Nix team on an experimental basis, as an experimental feature that is trivial enough to be approved with out requiring an RFC.
+It is still included here just to provide context.*
+
+In order to facilitate using `local-overlay` where the lower store is entirely read only (read only SQLite files too, not just store directory), it is useful to also implement a new "read-only" setting on the `local` store.
+The main thing this does is use SQLite's [immutable mode](https://www.sqlite.org/c3ref/open.html).
+
+This is a separate feature;
+it is perfectly possible to implement the `local-overlay` without this or vice-versa.
+But for maximum usability, we want to do both.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Because the `local-overlay` store is a completely separate store implementation, the interactions with the rest of Nix are fairly minimal and well-defined.
+In particular, users of other stores and not the `local-overlay` store will not be impacted at all.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Read-only local store is delicate
+
+SQLite's immutuable mode doesn't mean "we promise not to change the database".
```

Comment:
I found this paragraph very surprising. The rest of the RFC is written as if  the lower store can monotonically grow without being a problem for the overlays and that sounded really cool! And then suddenly this paragraph says that none of that is possible due to technical issues.

With this technical issue in place I am not so sure if a `local-overlay` store is that attractive or useful. It sounds like a deal-breaker that makes the abstraction kind of useless. If the sqlite database in the lower store can't be updated online, then what is the point?

How is it different from a lower store  containing `nix-registration-paths` text file with all valid paths that we `nix-store --load-db` on startup into the overlay sqlite database?   This is something we have already today and is well-tested (the ISO depend on it). And it seems to give the same benefits to me.     The lower store can be updated offline, and if you remount it then new paths can be added to the overlay sqlite database.

Reply:
Ah we address this later in https://github.com/NixOS/rfcs/pull/152/files#diff-36321d79e9a5897b43eaf0f9c622b4a14978b61641f934bc4d33cdc9a49f6669R345

I think we could put way more focus on  "This is an optimisation for `--load-db`".  As that is the main selling point here.